compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -2750,32 +2751,69 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
                             | LifetimeBinderKind::WhereBound
                     );
-                    let (span, sugg) = if span.is_empty() {
-                        let sugg = format!(
+
+                    let mut span = span;
+                    let mut rm_poly_trait_span = DUMMY_SP;
+                    let mut sugg = format!("{}, ", name.unwrap_or("'a"));
+
+                    if span.is_empty() {
+                        let mut generic_params = "".to_string();
+                        if let Some(with_poly_trait_ref) = self.with_poly_trait_ref.get(&node_id)
+                            && higher_ranked
+                        {
+                            generic_params = with_poly_trait_ref.generic_param_idents.iter().fold(
+                                "".to_string(),
+                                |mut generic_params, x| {
+                                    generic_params += x.as_str();
+                                    generic_params += ", ";
+                                    generic_params
+                                },
+                            );
+                            if !generic_params.is_empty() {
+                                rm_poly_trait_span = with_poly_trait_ref
+                                    .poly_trait
+                                    .with_hi(with_poly_trait_ref.trait_ref.lo());
+                            }
+                        }
+
+                        sugg = format!(
                             "{}<{}>{}",
                             if higher_ranked { "for" } else { "" },
-                            name.unwrap_or("'a"),
+                            format!("{}{}", generic_params, name.unwrap_or("'a")),
                             if higher_ranked { " " } else { "" },
                         );
-                        (span, sugg)
                     } else {
-                        let span = self
+                        span = self
                             .r
                             .tcx
                             .sess
                             .source_map()
                             .span_through_char(span, '<')
                             .shrink_to_hi();
-                        let sugg = format!("{}, ", name.unwrap_or("'a"));
-                        (span, sugg)
-                    };
+                    }
+
                     if higher_ranked {
                         let message = Cow::from(format!(
                             "consider making the {} lifetime-generic with a new `{}` lifetime",
                             kind.descr(),
                             name.unwrap_or("'a"),
                         ));
-                        should_continue = suggest(err, true, span, message, sugg);
+                        should_continue = if !rm_poly_trait_span.is_dummy() {
+                            // We should merge the higher-ranked lifetimes in following situation:
```

Comment:
Wording:

```suggestion
                            // We should merge binders in following situation:
```

Explanation: For me, “merging […] lifetimes” means replacing two previously distinct lifetimes with a single new one, so `'a, 'b => 'c`. Of course, it's relatively clear from context what's actually meant here but still I feel like we should use the more precise phrasing. Merging *binders* is `for<'a>, for<'b> => for<'a, 'b>`.

*Binder* is indeed jargon but so is higher-ranked lifetime.

Reply:
To make it more accessible we could clarify *binder* inline by saying `... binders (`for<>` param list) ...` or `... binders (list of higher-ranked generic params) ...`, dunno.